### ISSUE TITLE
Optimize Discrete data type by calming simultaneous events

### ISSUE THREAD

**Issue by HeinrichApfelmus** [HeinrichApfelmus/reactive-banana#19]
(This issue was kindly brought to my attention by Andrew Richards.)

Consider the following program:

```
import Reactive.Banana
import Debug.Trace

main = do
  (inputAH, inputTrigger) <- newAddHandler
  network <- compile $ do
    evt <- fromAddHandler inputAH
    let g = stepperD 0 evt
    let out = f <$> g <*> g
    reactimate $ print <$> changes out
  actuate network
  inputTrigger 3
  inputTrigger 4

f :: Int -> Int -> Int
f x y = trace ("f: " ++ show (x, y)) $ x + y
```

The function `f` is called more than once per update of the single input, even though it is unnecessary to call it more than once. Is it possible to optimize the `Discrete` type to avoid this?

Most likely yes, but I don't know how to perform this kind of optimization in the push-based implementation I currently have (version 0.4).

Basically, `Discrete` is implemented as a pair

```
type Discrete a = (Behavior a , Event a).
```

of a behavior (for recursion) and an update event. By duplicating `g`, the update event is also duplicated, resulting in two simultaneous update event occurrences for  the event `out`. What's currently missing from reactive-banana is a combinator

```
calm :: Event a -> Event a
```

that removes all but the last event occurrence from a bunch of simultaneous events. This would allow us to optimize the Discrete data type. The combinator seems to be safe from a semantic point of view.
